source/common/websocket/codec.h

```diff
@@ -0,0 +1,136 @@
+
+// Wire format (https://datatracker.ietf.org/doc/html/rfc6455#section-5.2)
+// of WebSocket frame:
+//
+//   0                   1                   2                   3
+//   0 1 2 3 4 5 6 7 8 9 0 1 2 3 4 5 6 7 8 9 0 1 2 3 4 5 6 7 8 9 0 1
+//  +-+-+-+-+-------+-+-------------+-------------------------------+
+//  |F|R|R|R| opcode|M| Payload len |    Extended payload length    |
+//  |I|S|S|S|  (4)  |A|     (7)     |             (16/64)           |
+//  |N|V|V|V|       |S|             |   (if payload len==126/127)   |
+//  | |1|2|3|       |K|             |                               |
+//  +-+-+-+-+-------+-+-------------+ - - - - - - - - - - - - - - - +
+//  |     Extended payload length continued, if payload len == 127  |
+//  + - - - - - - - - - - - - - - - +-------------------------------+
+//  |                               | Masking-key, if MASK set to 1 |
+//  +-------------------------------+-------------------------------+
+//  | Masking-key (continued)       |          Payload Data         |
+//  +-------------------------------- - - - - - - - - - - - - - - - +
+//  : .... Payload Data continued .... Payload Data continued ..... :
+//  + - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - +
+//  | .... Payload Data continued .... Payload Data continued ..... |
+//  +---------------------------------------------------------------+
+
+// In-memory representation of the contents of a WebSocket frame.
+struct Frame {
+  // Indicates that this is the final fragment in a message.
+  bool final_fragment_;
+  // Frame opcode.
+  uint8_t opcode_;
+  // The 4 byte fixed size masking key used to mask the payload. Masking/unmasking should be
+  // performed as described in https://datatracker.ietf.org/doc/html/rfc6455#section-5.3
+  absl::optional<uint32_t> masking_key_;
+  // Length of the payload as the number of bytes.
+  uint64_t payload_length_;
+  // WebSocket payload data (extension data and application data).
+  Buffer::InstancePtr payload_;
+};
+
+// Encoder encodes in memory WebSocket frames into frames in the wire format
+class Encoder : public Logger::Loggable<Logger::Id::websocket> {
+public:
+  Encoder() = default;
+
+  // Creates a new Websocket data frame header with the given frame data.
+  // @param frame supplies the frame to be encoded.
+  // @return std::vector<uint8_t> buffer with encoded header data.
+  absl::optional<std::vector<uint8_t>> encodeFrameHeader(const Frame& frame);
+};
+
+// Decoder decodes bytes in input buffer into in-memory WebSocket frames.
+class Decoder : public Logger::Loggable<Logger::Id::websocket> {
+public:
+  Decoder(uint64_t max_payload_length = kMaxPayloadBufferLength)
+      : max_payload_buffer_length_{std::min(max_payload_length, kMaxPayloadBufferLength)} {};
+  // Decodes the given buffer with WebSocket frame. Drains the input buffer when
+  // decoding succeeded (returns true). If the input is not sufficient to make a
```

Comment:
Those comment `Drains the input buffer when decoding succeeded (returns true)` should be updated now.